### PR TITLE
net: lib: dhcpv4_server: Add callback to allow user to set address

### PR DESCRIPTION
**Pull request by icsys-aal**
We needed functionality to set an ip address from the application. This functionality will allow you to create static leases.

Note that this callback is only triggered by a discover message, is that sufficient?

Let me know what you think!

### CODE REVIEW

**Comment by jukkar** [comment]
The code looks reasonable, but to get a proper review, please fix the compliance issue and make it a PR.

**Review by rlubos — COMMENTED** [review]
Just some nits.

File: include/zephyr/net/dhcpv4_server.h [hunk 1]
```diff
@@ -107,6 +107,35 @@ int net_dhcpv4_server_foreach_lease(struct net_if *iface,
...
+ * to the client. If the callback returns anything non-zero, the client will
+ * be assigned an address from the pool.
+ *
+ * @param iface Pointer to the network interface
+ * @param client_id Pointer to client requesting an address
+ * @param addr Address to be assigned to client
+ * @param user_data A valid pointer to user data or NULL
+ */
+typedef int (*net_dhcpv4_server_provider_cb_t)(struct net_if *iface,
+											   const struct dhcpv4_client_id *client_id,
```
Comment (rlubos) [inline include/zephyr/net/dhcpv4_server.h:126]: Something went wrong here and elsewhere, note that Zephyr uses 8 spaces/tab for indentation.

**Review by rlubos — DISMISSED** [review]

**Review by maass-hamburg — DISMISSED** [review]

**Comment by icsys-aal** [comment]
I have found a bug that I am currently fixing, please do not merge yet.

**Review by rlubos — APPROVED** [review]

**Review by maass-hamburg — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**